File: src/analytics/notify.py
"""Notification scheduling: admin template, user override, «smart» time.

Pure numbers and slots, like the rest of `src/analytics` — no ORM, no aiogram,
no wording. The «smart» time is an *observation* of when this person actually
records things, never a prescription of when they should eat
(`spec/notifications.md`, `spec/clinical.md`).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta

MINUTES_PER_DAY = 1440
#: how close to a slot a tick still counts as «now», minutes
SEND_WINDOW_MIN = 30
#: fewer records than this and there is no habit to speak of
MIN_EVENTS = 6
#: a cluster smaller than this is a one-off, not a habit
MIN_CLUSTER = 2
#: how far back the habit is read
WINDOW_DAYS = 21
#: default head start before the usual moment
LEAD_MIN = 15
#: two notifications closer than this are the same habit split in two
MIN_GAP_MIN = 45
#: slots when neither the profile nor the template says how many
DEFAULT_SLOTS = 3
MAX_SLOTS = 8

#: what a click on the notification opens
ACTIONS = ("camera", "text", "open")
#: template-level mode; a user picks one of these plus "default"
MODES = ("fixed", "smart", "off")
USER_MODES = ("default", "fixed", "smart", "off")
#: record kinds a template may learn from
SIGNALS = ("meal", "glucose", "activity", "weight", "wellbeing")


@dataclass(slots=True)
class Template:
    """Admin-authored notification — the default everyone starts from."""

    code: str
    title: str
    body: str
    action: str = "camera"
    mode: str = "fixed"
    times: tuple[str, ...] = ()
    signal: tuple[str, ...] = ()
    lead_min: int = LEAD_MIN
    enabled: bool = True
    sort: int = 0


@dataclass(slots=True)
class Pref:
    """What one person changed about one notification."""

    code: str
    mode: str = "default"
    times: tuple[str, ...] = ()


@dataclass(slots=True)
class Schedule:
    """Resolved plan for one notification and one person."""

    code: str
    mode: str
    times: tuple[str, ...] = field(default_factory=tuple)

    @property
    def active(self) -> bool:
        return self.mode != "off" and bool(self.times)


def parse_times(raw: str | None) -> tuple[str, ...]:
    """«8:00, 13:30,бред,25:00» -> ('08:00', '13:30'). Order kept, dupes dropped."""
    out: list[str] = []
    for chunk in str(raw or "").replace(";", ",").split(","):
        minute = parse_time(chunk)
        if minute is None:
            continue
        label = format_minute(minute)
        if label not in out:
            out.append(label)
    return tuple(out)


def parse_time(raw: str | None) -> int | None:
    """'8:05' -> 485. Anything that is not a time of day -> None."""
    text = str(raw or "").strip().replace(".", ":").replace("-", ":")
    if not text:
        return None
    head, _, tail = text.partition(":")
    if not head.isdigit():
        return None
    hour = int(head)
    minute = int(tail) if tail.isdigit() else 0
    if not (0 <= hour < 24 and 0 <= minute < 60):
        return None
    return hour * 60 + minute


def format_minute(minute: int) -> str:
    minute %= MINUTES_PER_DAY
    return f"{minute // 60:02d}:{minute % 60:02d}"


def minute_of_day(moment: datetime) -> int:
    return moment.hour * 60 + moment.minute


def smart_times(
    minutes: list[int],
    *,
    slots: int = DEFAULT_SLOTS,
    lead_min: int = LEAD_MIN,
    min_events: int = MIN_EVENTS,
    min_cluster: int = MIN_CLUSTER,
    min_gap: int = MIN_GAP_MIN,
) -> tuple[str, ...]:
    """Typical moments of a habit, shifted `lead_min` earlier.

    One-dimensional k-means over minutes of day: quantile seeds, at most 25
    passes, median (not mean) per cluster so a single 2 a.m. record cannot drag
    a slot across the evening. Clusters closer than `min_gap` are then merged —
    one outlier can split a habit in two, and two notifications five minutes
    apart are noise, not a schedule. Too few records — empty result, and the
    caller falls back to the template: inventing a habit out of three points
    would be a guess dressed as a measurement.
    """
    points = sorted(m % MINUTES_PER_DAY for m in minutes if m is not None)
    slots = max(1, min(int(slots or DEFAULT_SLOTS), MAX_SLOTS))
    if len(points) < max(min_events, 1):
        return ()
    if len(points) < slots:
        slots = len(points)

    centers = _quantile_seeds(points, slots)
    assignment: list[list[int]] = []
    for _ in range(25):
        buckets: list[list[int]] = [[] for _ in centers]
        for point in points:
            best = min(range(len(centers)), key=lambda i: abs(point - centers[i]))
            buckets[best].append(point)
        moved = False
        for i, bucket in enumerate(buckets):
            if not bucket:
                continue
            center = _median(bucket)
            if center != centers[i]:
                centers[i] = center
                moved = True
        assignment = buckets
        if not moved:
            break

    found: list[tuple[int, int]] = []   # (slot minute, cluster size)
    for bucket in assignment:
        if len(bucket) < min_cluster:
            continue
        target = (_median(bucket) - max(0, lead_min)) % MINUTES_PER_DAY
        found.append((target - target % 5, len(bucket)))
    return tuple(format_minute(minute) for minute in _merge_close(found, min_gap))


def _merge_close(found: list[tuple[int, int]], min_gap: int) -> list[int]:
    """Collapse neighbouring slots; the bigger cluster keeps its time."""
    out: list[tuple[int, int]] = []
    for minute, size in sorted(found):
        if out and minute - out[-1][0] < max(1, min_gap):
            previous, previous_size = out[-1]
            keep = previous if previous_size >= size else minute
            out[-1] = (keep, previous_size + size)
            continue
        out.append((minute, size))
    return [minute for minute, _size in out]


def _quantile_seeds(points: list[int], k: int) -> list[int]:
    """Deterministic seeds: the k inner quantiles of the sample."""
    n = len(points)
    return [points[min(n - 1, int((i + 0.5) * n / k))] for i in range(k)]


def _median(values: list[int]) -> int:
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) // 2


def recent_minutes(
    moments: list[datetime], *, now: datetime, window_days: int = WINDOW_DAYS
) -> list[int]:
    """Local timestamps -> minutes of day, keeping only the recent window."""
    edge = now - timedelta(days=window_days)
    return [minute_of_day(m) for m in moments if m >= edge]


def resolve(
    template: Template,
    pref: Pref | None = None,
    *,
    smart: tuple[str, ...] = (),
) -> Schedule:
    """Admin template first, the person's own choice on top of it.

    A template switched off is off for everyone: `enabled=0` is the operator
    removing the notification, not a default to be overridden.
    """
    if not template.enabled:
        return Schedule(template.code, "off", ())
    mode = pref.mode if pref is not None else "default"
    if mode not in USER_MODES:
        mode = "default"
    if mode == "default":
        mode = template.mode if template.mode in MODES else "fixed"
        times = template.times
    elif mode == "off":
        return Schedule(template.code, "off", ())
    elif mode == "fixed":
        times = (pref.times if pref else ()) or template.times
    else:  # smart
        times = smart or template.times
    if mode == "smart" and not smart:
        # No habit yet — the template's times stand in, and the mode says so.
        return Schedule(template.code, "smart", template.times)
    return Schedule(template.code, mode, tuple(times))


def due_slots(
    schedule: Schedule,
    *,
    local_now: datetime,
    window_min: int = SEND_WINDOW_MIN,
) -> tuple[str, ...]:
    """Which slots of this schedule the current tick should deliver.

    A slot stays due for `window_min` minutes: a cron that runs every quarter
    of an hour must not drop a notification it happened to step over.
    """
    if not schedule.active:
        return ()
    now = minute_of_day(local_now)
    out: list[str] = []
    for label in schedule.times:
        slot = parse_time(label)
        if slot is None:
            continue
        delta = (now - slot) % MINUTES_PER_DAY
        if delta < window_min:
            out.append(format_minute(slot))
    return tuple(out)


__all__ = [
    "ACTIONS",
    "DEFAULT_SLOTS",
    "LEAD_MIN",
    "MIN_CLUSTER",
    "MIN_EVENTS",
    "MODES",
    "SEND_WINDOW_MIN",
    "SIGNALS",
    "USER_MODES",
    "WINDOW_DAYS",
    "Pref",
    "Schedule",
    "Template",
    "due_slots",
    "format_minute",
    "minute_of_day",
    "parse_time",
    "parse_times",
    "recent_minutes",
    "resolve",
    "smart_times",
]

File: web/app/js/notify.js
/**
 * notify.js — notifications on the device side: merge the admin's list with
 * this person's own choices, compute the «smart» time from local records, and
 * send the resulting schedule (times only, never records) to the server.
 * Mirrors src/analytics/notify.py — same clustering, same thresholds.
 * spec: spec/notifications.md.
 */
import { API_BASE } from './config.js';
import { getKV, setKV, getAll } from './db.js';

const MINUTES_PER_DAY = 1440;
export const MIN_EVENTS = 6;      // fewer records than this: no habit to read
export const MIN_CLUSTER = 2;     // a one-off is not a daily notification
export const WINDOW_DAYS = 21;
export const DEFAULT_SLOTS = 3;
export const MIN_GAP_MIN = 45;    // closer than this is one habit split in two
const MAX_SLOTS = 8;

/**
 * signal name -> IndexedDB store it is recorded in. `activity` has no store
 * here: the web MVP records no workouts (they stay in the bot), so a template
 * whose only signal is activity keeps the admin's times.
 */
const SIGNAL_STORES = {
    meal: 'meals',
    glucose: 'glucose',
    weight: 'weight',
    wellbeing: 'wellbeing',
};

export function formatMinute(minute) {
    const m = ((minute % MINUTES_PER_DAY) + MINUTES_PER_DAY) % MINUTES_PER_DAY;
    return String(Math.floor(m / 60)).padStart(2, '0') + ':' + String(m % 60).padStart(2, '0');
}

export function parseTime(raw) {
    const text = String(raw || '').trim().replace(/[.-]/g, ':');
    if (!text) return null;
    const [head, tail] = text.split(':');
    if (!/^\d+$/.test(head || '')) return null;
    const hour = parseInt(head, 10);
    const minute = /^\d+$/.test(tail || '') ? parseInt(tail, 10) : 0;
    if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
    return hour * 60 + minute;
}

export function parseTimes(raw) {
    const out = [];
    for (const chunk of String(raw || '').split(/[,;]/)) {
        const minute = parseTime(chunk);
        if (minute === null) continue;
        const label = formatMinute(minute);
        if (!out.includes(label)) out.push(label);
    }
    return out;
}

function median(values) {
    const ordered = [...values].sort((a, b) => a - b);
    const mid = Math.floor(ordered.length / 2);
    return ordered.length % 2 ? ordered[mid] : Math.floor((ordered[mid - 1] + ordered[mid]) / 2);
}

/**
 * Typical moments of a habit, shifted `leadMin` earlier: one-dimensional
 * k-means over minutes of day with quantile seeds and a median per cluster, so
 * one 2 a.m. record cannot drag a slot across the evening; clusters closer than
 * `minGap` are then merged. Too few records — empty result, and the caller
 * falls back to the admin's times. Mirrors src/analytics/notify.py.
 */
export function smartTimes(minutes, { slots = DEFAULT_SLOTS, leadMin = 15, minEvents = MIN_EVENTS, minCluster = MIN_CLUSTER, minGap = MIN_GAP_MIN } = {}) {
    const points = minutes
        .filter((m) => Number.isFinite(m))
        .map((m) => ((m % MINUTES_PER_DAY) + MINUTES_PER_DAY) % MINUTES_PER_DAY)
        .sort((a, b) => a - b);
    let k = Math.max(1, Math.min(Math.round(slots) || DEFAULT_SLOTS, MAX_SLOTS));
    if (points.length < Math.max(minEvents, 1)) return [];
    if (points.length < k) k = points.length;

    const centers = [];
    for (let i = 0; i < k; i++) centers.push(points[Math.min(points.length - 1, Math.floor((i + 0.5) * points.length / k))]);

    let buckets = [];
    for (let pass = 0; pass < 25; pass++) {
        buckets = centers.map(() => []);
        for (const point of points) {
            let best = 0;
            for (let i = 1; i < centers.length; i++) {
                if (Math.abs(point - centers[i]) < Math.abs(point - centers[best])) best = i;
            }
            buckets[best].push(point);
        }
        let moved = false;
        buckets.forEach((bucket, i) => {
            if (!bucket.length) return;
            const center = median(bucket);
            if (center !== centers[i]) { centers[i] = center; moved = true; }
        });
        if (!moved) break;
    }

    const found = [];   // [slot minute, cluster size]
    for (const bucket of buckets) {
        if (bucket.length < minCluster) continue;
        const target = ((median(bucket) - Math.max(0, leadMin)) % MINUTES_PER_DAY + MINUTES_PER_DAY) % MINUTES_PER_DAY;
        found.push([target - (target % 5), bucket.length]);
    }
    return mergeClose(found, minGap).map(formatMinute);
}

/** Collapse neighbouring slots; the bigger cluster keeps its time. */
function mergeClose(found, minGap) {
    const out = [];
    for (const [minute, size] of found.sort((a, b) => a[0] - b[0])) {
        const last = out[out.length - 1];
        if (last && minute - last[0] < Math.max(1, minGap)) {
            out[out.length - 1] = [last[1] >= size ? last[0] : minute, last[1] + size];
            continue;
        }
        out.push([minute, size]);
    }
    return out.map(([minute]) => minute);
}

/** Minutes of day of everything recorded for these signals in the last weeks. */
export async function signalMinutes(signals, { windowDays = WINDOW_DAYS } = {}) {
    const edge = Date.now() - windowDays * 86400000;
    const out = [];
    for (const signal of signals) {
        const store = SIGNAL_STORES[signal];
        if (!store) continue;
        let rows = [];
        try {
            rows = await getAll(store);
        } catch (e) {
            continue;   // store may not exist in an older database
        }
        for (const row of rows) {
            const ts = new Date(row.ts).getTime();
            if (!Number.isFinite(ts) || ts < edge) continue;
            const local = new Date(ts);
            out.push(local.getHours() * 60 + local.getMinutes());
        }
    }
    return out;
}

/** How many notifications a day this signal deserves. */
async function slotsFor(notification) {
    if (notification.signal.includes('meal')) {
        const profile = await getKV('profile', {});
        const perDay = parseInt(profile.mealsPerDay, 10);
        if (Number.isFinite(perDay) && perDay > 0) return perDay;
    }
    return notification.adminTimes.length || DEFAULT_SLOTS;
}

/** Smart times for one notification, or [] when there is no habit yet. */
export async function computeSmart(notification) {
    const minutes = await signalMinutes(notification.signal);
    return smartTimes(minutes, {
        slots: await slotsFor(notification),
        leadMin: notification.leadMin,
    });
}

export function tzOffsetMinutes() {
    return -new Date().getTimezoneOffset();
}

/**
 * The merged list: what the admin set, what this device changed, and what
 * «smart» currently resolves to. The server copy is authoritative for the
 * admin's half; the local copy answers while offline.
 */
export async function loadNotifications(clientId) {
    let list = await getKV('notifications', []);
    try {
        const res = await fetch(API_BASE + 'notify_config.php?clientId=' + encodeURIComponent(clientId || ''));
        const data = await res.json();
        if (Array.isArray(data.notifications)) {
            list = data.notifications;
            await setKV('notifications', list);
        }
    } catch (e) { /* offline: the cached list still renders */ }
    for (const item of list) {
        item.smartTimes = item.mode === 'smart' || item.adminMode === 'smart' ? await computeSmart(item) : [];
    }
    return list;
}

/** Times this notification will actually fire at, in the order of priority. */
export function effectiveTimes(item) {
    if (item.mode === 'off') return [];
    if (item.mode === 'default') return item.adminMode === 'off' ? [] : item.adminTimes;
    if (item.mode === 'fixed') return item.times.length ? item.times : item.adminTimes;
    return item.smartTimes && item.smartTimes.length ? item.smartTimes : item.adminTimes;
}

/** Push the device's choices (and the resolved smart times) to the server. */
export async function saveNotifications(clientId, list) {
    await setKV('notifications', list);
    const prefs = list.map((item) => ({
        code: item.code,
        mode: item.mode,
        times: item.mode === 'smart' ? (item.smartTimes || []) : item.times,
    }));
    try {
        await fetch(API_BASE + 'notify_config.php', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ clientId, tzOffset: tzOffsetMinutes(), prefs }),
        });
    } catch (e) {
        return false;   // saved locally; the next open retries
    }
    return true;
}
